### PR TITLE
test: reproduce rustfix panic on overlapping suggestions

### PR DESCRIPTION
**Pull request by raushan728**
Relates to #13030.

This PR adds an end-to-end test in `tests/testsuite/fix_n_times.rs` to reproduce and track the current panic that occurs when `rustc` emits multiple overlapping `MachineApplicable` suggestions for the exact same span.

*(Note: PR scoped down to just the test case so we can track this behavior while the actual fix/design is being discussed).*

### CODE REVIEW

File: crates/rustfix/src/lib.rs [hunk 1]
```diff
@@ -281,12 +281,63 @@ impl CodeFix {
...
+            // See rust-lang/cargo#13030.
+            Err(Error::AlreadyReplaced { .. }) => continue,
+            Err(e) => return Err(e),
+        }
     }
     fix.finish()
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (epage) [inline crates/rustfix/src/lib.rs]: Our [contrib guide](https://doc.crates.io/contrib/process/working-on-cargo.html#submitting-a-pull-request) asks for tests to be added in a prior commit, reproducing the current problem (with the tests passing).  The fix would then update the tests to show the new behavior

Comment (raushan728) [reply crates/rustfix/src/lib.rs]: @epage done u can review it now

Comment (epage) [reply crates/rustfix/src/lib.rs]: Note I was just focusing on process.  As there isn't an end-to-end test because this is fixing a theoretical problem and not a real problem, I don't feel qualified to review this.

**Comment by epage** [comment]
r? ehuss

Would you be up for this considering this is a fix for a theoretical problem?  I don't have much context for the actual problem to be able to tell if this addresses it.

**Comment by epage** [comment]
From https://github.com/rust-lang/cargo/issues/13030#issuecomment-1822033539

> I'd leave it open. Even though these types of suggestions aren't currently supported, it probably shouldn't panic.

It requires a multi-suggestion that is Machine Applicable but because we can't support that (at the UX level), the compiler shouldn't emit it.

**Comment by raushan728** [comment]
@weihanglo @epage about the e2e test making a real-world snapshot test is pretty hard bcz we'd have to force rustc to emit multiple exclusive `MachineApplicable` suggestions for the exact same span, which it shouldn't really do anyway. 

that's exactly y i added a unit test to mock this edge case directly in `apply_suggestions`. the main goal is just defensive: stop the panic and gracefully skip overlaps like ehuss suggested in the original issue

let me know what u guys think or if u want me to try a diff approach

**Comment by epage** [comment]
We do have a mock rustc for cargo's test suite in https://github.com/rust-lang/cargo/blob/master/tests/testsuite/fix_n_times.rs

**Comment by raushan728** [comment]
hi @epage 
thanks for the pointer to `rustc-fix-shim` — exactly what I needed

added an e2e test in `tests/testsuite/fix_n_times.rs` (`Step::TwoFixExclusive`) — forces the mock compiler to emit a single diagnostic with multiple overlapping `MachineApplicable` suggestions on the same span.

while digging in, realized the real issue was in `CodeFix::apply`, not just `apply_suggestions` — it was propagating `AlreadyReplaced` for intra-suggestion conflicts. Fixed it: if a conflict occurs but `applied_any` is true, we already committed a solution for that diagnostic, so we now gracefully skip the conflicting ones instead of panicking.

LMK what you think!

**Comment by epage** [comment]
@weihanglo 

> Is it possible to have the test case from a real-world diagnostic? If not, I slightly lean towards we leave it.

A benefit to the panic that came up during office hours is that *if* the compiler developer tests their fix, they will see the panic and know that they shouldn't do this.  If we silently ignore it, there will be no feedback look for the compiler developer.

**Comment by weihanglo** [comment]
Yes. I also want to "leave" panic there. I don't think without a real world example this is going to anywhere.

**Comment by raushan728** [comment]
@epage @weihanglo Makes total sense! I get the need for a loud feedback loop. 

Quick thought before I close this: Should we catch this specific overlap and throw a targeted panic instead of the generic one? 

e.g., `panic!("Compiler Bug: Emitted multiple overlapping MachineApplicable suggestions for the exact same span.")`

This keeps the panic but tells the compiler dev exactly what went wrong. Let me know if u want this

**Comment by epage** [comment]
I could see it being helpful to have a more specific panic or error message.  In particular, Cargo has an `InternalError` that prints the message and tells the user to report the issue to us.

**Comment by raushan728** [comment]
@epage updated the PR It now returns an `InternalError` with the specific diagnostic message instead of skipping the overlap along with the expected output in the e2e test. Let me know if it looks good now

File: tests/testsuite/fix_n_times.rs [hunk 2]
```diff
@@ -521,3 +632,21 @@ rustc fix shim error count=2
         "// fix-count 1",
     );
 }
+
+#[cargo_test]
+fn fix_exclusive_suggestions() {
```
Comment (epage) [inline tests/testsuite/fix_n_times.rs:588]: Tests should be in commits before the change

**Comment by epage** [comment]
Please clean up the commits to reflect how this should be reviewed and merged and not how this was developed.

**Comment by raushan728** [comment]
> Please clean up the commits to reflect how this should be reviewed and merged and not how this was developed.

There are now exactly two commits:
1. **Commit 1:** Introduces the E2E test demonstrating the original generic `AlreadyReplaced` panic behavior.
2. **Commit 2:** Implements the logic to treat intra-suggestion overlaps as an `InternalError` and updates the test expectations accordingly.

File: crates/rustfix/src/error.rs [hunk 3]
```diff
@@ -11,7 +11,6 @@ pub enum Error {
     #[error("invalid range {0:?}, original data is only {1} byte long")]
     DataLengthExceeded(Range<usize>, usize),
 
-    #[non_exhaustive]
```
Comment (epage) [inline crates/rustfix/src/error.rs:14]: Why was this removed?

File: crates/rustfix/src/lib.rs [hunk 4]
```diff
@@ -232,16 +232,49 @@ impl CodeFix {
...
+                    }
+                }
             }
+            // Commit this solution so subsequent solutions see it as committed.
+            self.data.commit();
+            applied_any = true;
+        }
+        if applied_any {
+            self.modified = true;
         }
```
Comment (epage) [inline crates/rustfix/src/lib.rs]: Why was this changed?

File: crates/rustfix/src/error.rs (hunk 3 above)
Comment (raushan728) [reply crates/rustfix/src/error.rs:14]: Accidentally removed while restructuring restoring it now

File: crates/rustfix/src/lib.rs (hunk 4 above)
Comment (raushan728) [reply crates/rustfix/src/lib.rs]: Since the new error path does an early return `Err(...)`, I needed a way to track whether any solution was successfully committed before that point. `applied_any` handles that — `self.modified` is only set if at least one solution actually committed without conflict

File: crates/rustfix/src/error.rs [hunk 5]
```diff
@@ -10,7 +10,6 @@ pub enum Error {
 
     #[error("invalid range {0:?}, original data is only {1} byte long")]
     DataLengthExceeded(Range<usize>, usize),
-
```
Comment (epage) [inline crates/rustfix/src/error.rs:13]: Now there is a spurious blank line removal

File: crates/rustfix/src/lib.rs (hunk 4 above)
Comment (epage) [reply crates/rustfix/src/lib.rs]: I think I'm failing to see why adding an early return requires changing the code in the situation when no early return is taken

File: crates/rustfix/src/error.rs (hunk 5 above)
Comment (raushan728) [reply crates/rustfix/src/error.rs:13]: I just restored the `#[non_exhaustive]` macro and ran `cargo fmt`, which formatted it this way automatically. Should I manually force a blank line here just to keep the diff clean?

Comment (epage) [reply crates/rustfix/src/error.rs:13]: Most other variants have a line between them and `cargo fmt` is leavign them untouched.  I also tried to reproduce this and am not seeing it.

File: crates/rustfix/src/lib.rs (hunk 4 above)
Comment (raushan728) [reply crates/rustfix/src/lib.rs]: I just ended up over-complicating it by trying to refactor the loop earlier. I've completely reverted those so it strictly just adds the early return

File: tests/testsuite/fix_n_times.rs [hunk 6]
```diff
@@ -472,3 +632,41 @@ rustc fix shim error count=2
...
+        },
+        str![[r#"
+[CHECKING] foo v0.0.1 ([ROOT]/foo)
+[WARNING] error applying suggestions to `src/lib.rs`
+
+The full error message was:
+
+> cannot replace slice of data that was already replaced
+
+[..]
```
Comment (epage) [inline tests/testsuite/fix_n_times.rs]: What is this redacting?

File: tests/testsuite/fix_n_times.rs [hunk 7]
```diff
@@ -636,35 +636,24 @@ rustc fix shim error count=2
...
-https://github.com/rust-lang/rust/issues
-quoting the full output of this command we'd be very appreciative!
-Note that you may be able to make some more progress in the near-term
-fixing code with the `--broken-code` flag
-
-[FIXED] src/lib.rs (0 fixes)
-rustc fix shim exclusive comment 1
-[WARNING] `foo` (lib) generated 1 warning (run `cargo fix --lib -p foo` to apply 1 suggestion)
-[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+[ERROR] rustc emitted multiple overlapping MachineApplicable suggestions for the exact same span.
```
Comment (epage) [inline tests/testsuite/fix_n_times.rs]: So before we were talking about this being a panic and wanting to avoid that.  Instead, this seems to be a rustc ICE turned into a warning.  We are now turning that warning into an error.  I'm wondering which is more appropriate.

@weihanglo any thoughts?

Comment (weihanglo) [reply tests/testsuite/fix_n_times.rs]: If there are multiple build targets, would it fixes other targets but the warned one with the old logic? If yes, we probably want the `--keep-going`-like behavior. If the overall goal here is letting rustc dev know the ICE, hard error seems also fine. We are fixing a theoretical issue so I don't have strong preference.

This also reminds me we should have some different strategy in `cargo-fix` and `cargo-fixit` to maximize fix applications (and also the interactive application).

File: tests/testsuite/fix_n_times.rs (hunk 6 above)
Comment (raushan728) [reply tests/testsuite/fix_n_times.rs]: In the first commit, it was just redacting the standard long boilerplate text that usually follows the error (the 'This likely indicates a bug in either rustc or cargo itself...' paragraph) to keep the snapshot concise.

However, this [..] is no longer present in the latest updated snapshot

File: tests/testsuite/fix_n_times.rs (hunk 7 above)
Comment (epage) [reply tests/testsuite/fix_n_times.rs]: I'm then leaning towards scoping this PR down to just adding tests for the case so we can be aware of and track this behavior.

Comment (raushan728) [reply tests/testsuite/fix_n_times.rs]: So, should I drop commit 2?

Comment (epage) [reply tests/testsuite/fix_n_times.rs]: Yes, sorry about that and thank you for all of the work you have put into this as we've better understood the problem and worked to resolve this.

Comment (raushan728) [reply tests/testsuite/fix_n_times.rs]: No worries at all just to be sure before I update the PR: is the first commit (the test reproducing the panic) good to go exactly as it is right now? If yes, I will go ahead, drop the second commit

**Comment by raushan728** [comment]
Hi @epage i have dropped the second commit. plz take look once u get a chance!

**Comment by epage** [comment]
Could you remove the merge commit?

**Comment by raushan728** [comment]
> Could you remove the merge commit?

Sorry for that i have removed it now
